linuxUtils/linux.go

```diff
@@ -383,3 +389,99 @@ func BinfmtMiscNamespacingSupported() (bool, error) {
 
 	return cmp >= 0, nil
 }
+
+// ShmSysctlUserNamespaced checks if /proc/sys/kernel/shm* sysctls can be
+// written to from within a user-ns empirically (by running an experiment).
+// Support for such namespacing was added in upstream kernel 6.9 (commit
+// 50ec499b9a43) but other distros (e.g., Ubuntu) may have backported to earlier
+// versions so just checking for the kernel version is not sufficient.
+func ShmSysctlUserNamespaced() (bool, error) {
+	logrus.Debugf("Running /proc/sys/kernel/shm* sysctl namespacing check on host.")
+
+	currCwd, err := os.Getwd()
+	if err != nil {
+		return false, err
+	}
+
+	// This is the function that will check if /proc/sys/kernel/shm* sysctls are
+	// namespaced and can be written to.
+	execFunc := func() {
+		logrus.Debugf("- shm ns check: execFunc: lock OS thread")
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
+
+		// Make ourselves root within the user ns
+		logrus.Debugf("- shm ns check: execFunc: setresuid")
+		if err := setxid.Setresuid(0, 0, 0); err != nil {
+			logrus.Debugf("- shiftfs check: execFunc: failed: %v", err)
+			os.Exit(1)
+		}
+		logrus.Debugf("- shm ns check: execFunc: setresgid")
+		if err := setxid.Setresgid(0, 0, 0); err != nil {
+			logrus.Debugf("- shm ns check: execFunc: failed: %v", err)
+			os.Exit(1)
+		}
+
+		// Try opening the /proc/sys/kernel/shm* files for read-write access;
+		// failure to do so means the sysctl is not accesible from a user-ns.
+		sysctls := []string{
+			"/proc/sys/kernel/shmmni",
+			"/proc/sys/kernel/shmmax",
+			"/proc/sys/kernel/shmall",
+		}
+		for _, sysctl := range sysctls {
+			f, err := os.OpenFile(sysctl, os.O_RDWR, 0o644)
+			if err != nil {
+				if errors.Is(err, syscall.EPERM) {
+					logrus.Debugf("- shm ns check: execFunc: sysctl %s is not writeable from user-ns (EPERM)", sysctl)
+				} else {
```

Comment:
Are all three of these nodes required for these tests to succeed? IOW, don't you want to continue the test if one of these is not present in a given kernel version? e.g., `if err.IsnotExist() { continue }`

Reply:
I think it's best to test for all 3 sysctls just to be sure; but in reality they come as a group, so either they will all pass or all fail, so whether we fail on the first or continue won't really matter.